File: agent-governance-golang/packages/agentmesh/sandbox.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package agentmesh

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"os/exec"
	"regexp"
	"strings"
	"sync"
	"time"
)

// randomHex returns 2*n hex characters drawn from crypto/rand.
//
// A failure from crypto/rand indicates a broken entropy source (e.g.
// /dev/urandom is unreachable) — the sandbox cannot safely produce
// unique session identifiers in that state, so we panic rather than
// return a degraded fallback. This is the same convention Go uses for
// `rand.Read` in security-critical contexts (`crypto/rsa.GenerateKey`
// etc.).
func randomHex(n int) string {
	b := make([]byte, n)
	if _, err := rand.Read(b); err != nil {
		panic(fmt.Sprintf("sandbox: crypto/rand failed: %v", err))
	}
	return hex.EncodeToString(b)
}

// Docker subprocess deadlines. ``docker exec`` runs user-supplied code so
// it falls back to SandboxConfig.TimeoutSeconds when set; the others are
// short-lived control-plane operations against the local Docker daemon.
const (
	dockerInfoTimeout    = 5 * time.Second
	dockerStartTimeout   = 30 * time.Second
	dockerExecTimeout    = 60 * time.Second
	dockerRemoveTimeout  = 15 * time.Second
)

// SessionStatus represents the lifecycle state of a sandbox session.
type SessionStatus string

const (
	SessionCreating  SessionStatus = "creating"
	SessionRunning   SessionStatus = "running"
	SessionStopped   SessionStatus = "stopped"
	SessionDestroyed SessionStatus = "destroyed"
	SessionError     SessionStatus = "error"
)

// ExecutionStatus represents the state of a code execution request.
type ExecutionStatus string

const (
	ExecutionPending   ExecutionStatus = "pending"
	ExecutionRunning   ExecutionStatus = "running"
	ExecutionCompleted ExecutionStatus = "completed"
	ExecutionFailed    ExecutionStatus = "failed"
	ExecutionTimeout   ExecutionStatus = "timeout"
	ExecutionKilled    ExecutionStatus = "killed"
)

// SandboxConfig holds resource limits and security settings for a sandbox session.
type SandboxConfig struct {
	TimeoutSeconds float64           `json:"timeout_seconds"`
	MemoryMB       int               `json:"memory_mb"`
	CPULimit       float64           `json:"cpu_limit"`
	NetworkEnabled bool              `json:"network_enabled"`
	ReadOnlyFS     bool              `json:"read_only_fs"`
	EnvVars        map[string]string `json:"env_vars"`
}

// SandboxResult holds the output and metadata from a sandbox code execution.
type SandboxResult struct {
	Success         bool    `json:"success"`
	ExitCode        int     `json:"exit_code"`
	Stdout          string  `json:"stdout"`
	Stderr          string  `json:"stderr"`
	DurationSeconds float64 `json:"duration_seconds"`
	Killed          bool    `json:"killed"`
	KillReason      string  `json:"kill_reason"`
}

// SessionHandle represents an active sandbox session.
type SessionHandle struct {
	AgentID   string        `json:"agent_id"`
	SessionID string        `json:"session_id"`
	Status    SessionStatus `json:"status"`
}

// ExecutionHandle represents a code execution within a sandbox session.
type ExecutionHandle struct {
	ExecutionID string          `json:"execution_id"`
	AgentID     string          `json:"agent_id"`
	SessionID   string          `json:"session_id"`
	Status      ExecutionStatus `json:"status"`
	Result      *SandboxResult  `json:"result"`
}

// DefaultSandboxConfig returns a SandboxConfig with sensible defaults.
func DefaultSandboxConfig() *SandboxConfig {
	return &SandboxConfig{
		TimeoutSeconds: 60,
		MemoryMB:       512,
		CPULimit:       1.0,
		NetworkEnabled: false,
		ReadOnlyFS:     true,
		EnvVars:        make(map[string]string),
	}
}

// SandboxProvider defines the interface for managing isolated code execution environments.
type SandboxProvider interface {
	CreateSession(agentID string, config *SandboxConfig) (*SessionHandle, error)
	ExecuteCode(agentID, sessionID, code string) (*ExecutionHandle, error)
	DestroySession(agentID, sessionID string) error
	IsAvailable() bool
}

// DockerSandboxProvider implements SandboxProvider using the Docker CLI.
type DockerSandboxProvider struct {
	image          string
	containers     map[string]string // key: "agentID:sessionID", value: container name
	mu             sync.Mutex
	availableOnce  sync.Once
	available      bool
}

// NewDockerSandboxProvider creates a DockerSandboxProvider with the given base image.
//
// The constructor is non-blocking: Docker daemon reachability is probed
// lazily on the first call to [DockerSandboxProvider.IsAvailable] (or
// equivalently on the first [DockerSandboxProvider.CreateSession]).
// Callers that never invoke a sandbox method — for example, unit tests
// that exercise other code paths — never pay the `docker info` cost.
func NewDockerSandboxProvider(image string) *DockerSandboxProvider {
	return &DockerSandboxProvider{
		image:      image,
		containers: make(map[string]string),
	}
}

// IsAvailable reports whether the Docker daemon is reachable. The
// underlying `docker info` probe runs at most once per provider; the
// result is cached for the lifetime of the receiver. A provider whose
// first probe fails will not retry — construct a new provider if the
// daemon becomes available later.
func (p *DockerSandboxProvider) IsAvailable() bool {
	p.availableOnce.Do(func() {
		ctx, cancel := context.WithTimeout(context.Background(), dockerInfoTimeout)
		defer cancel()
		if err := exec.CommandContext(ctx, "docker", "info").Run(); err == nil {
			p.available = true
		}
	})
	return p.available
}

func containerKey(agentID, sessionID string) string {
	return agentID + ":" + sessionID
}

func containerName(agentID, sessionID string) string {
	return fmt.Sprintf("agt-sandbox-%s-%s", agentID, sessionID)
}

// containerIDPattern is the character set Docker accepts in `--name`
// values (alphanumeric, `_`, `.`, `-`). agentID values that contain
// anything outside this set get rejected at the sandbox-provider
// boundary so they cannot land in a `docker run --name` argument where
// Docker would either reject them, truncate them, or — for a few
// characters like `$` — produce a name whose collisions cause
// surprising race conditions across concurrent sessions.
var containerIDPattern = regexp.MustCompile(`^[a-zA-Z0-9_.-]+$`)

// CreateSession starts a hardened Docker container for the given agent.
func (p *DockerSandboxProvider) CreateSession(agentID string, config *SandboxConfig) (*SessionHandle, error) {
	if !p.IsAvailable() {
		return nil, fmt.Errorf("docker is not available")
	}
	if !containerIDPattern.MatchString(agentID) {
		return nil, fmt.Errorf("invalid agentID %q: must match %s", agentID, containerIDPattern)
	}
	if config == nil {
		config = DefaultSandboxConfig()
	}

	// `time.Now().UnixNano()` alone collides on low-resolution clocks
	// (Windows commonly resolves to ~100ns) and on closely-spaced
	// concurrent calls; append a crypto/rand suffix so two sessions
	// created in the same nanosecond cannot map to the same container.
	sessionID := fmt.Sprintf("%d-%s", time.Now().UnixNano(), randomHex(8))
	name := containerName(agentID, sessionID)

	args := []string{
		"run", "-d",
		"--name", name,
		fmt.Sprintf("--memory=%dm", config.MemoryMB),
		fmt.Sprintf("--cpus=%.2f", config.CPULimit),
		"--cap-drop", "ALL",
		"--security-opt", "no-new-privileges",
	}

	if config.ReadOnlyFS {
		args = append(args, "--read-only")
	}
	if !config.NetworkEnabled {
		args = append(args, "--network", "none")
	}
	for k, v := range config.EnvVars {
		args = append(args, "-e", fmt.Sprintf("%s=%s", k, v))
	}

	args = append(args, p.image, "sleep", "infinity")

	ctx, cancel := context.WithTimeout(context.Background(), dockerStartTimeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, "docker", args...).CombinedOutput()
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			return nil, fmt.Errorf("docker run timed out after %s", dockerStartTimeout)
		}
		return nil, fmt.Errorf("docker run failed: %w: %s", err, strings.TrimSpace(string(out)))
	}

	p.mu.Lock()
	if p.containers == nil {
		p.containers = make(map[string]string)
	}
	p.containers[containerKey(agentID, sessionID)] = name
	p.mu.Unlock()

	return &SessionHandle{
		AgentID:   agentID,
		SessionID: sessionID,
		Status:    SessionRunning,
	}, nil
}

// ExecuteCode runs a command inside an existing sandbox session container.
func (p *DockerSandboxProvider) ExecuteCode(agentID, sessionID, code string) (*ExecutionHandle, error) {
	p.mu.Lock()
	name, ok := p.containers[containerKey(agentID, sessionID)]
	p.mu.Unlock()
	if !ok {
		return nil, fmt.Errorf("session %s:%s not found", agentID, sessionID)
	}

	// Same collision concern as sessionID — keep the nanosecond prefix
	// for human-readable ordering and append a random suffix.
	execID := fmt.Sprintf("exec-%d-%s", time.Now().UnixNano(), randomHex(4))
	start := time.Now()

	ctx, cancel := context.WithTimeout(context.Background(), dockerExecTimeout)
	defer cancel()
	cmd := exec.CommandContext(ctx, "docker", "exec", name, "sh", "-c", code)
	var stdout, stderr strings.Builder
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	duration := time.Since(start).Seconds()

	exitCode := 0
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			return nil, fmt.Errorf("docker exec timed out after %s", dockerExecTimeout)
		}
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			return nil, fmt.Errorf("docker exec failed: %w", err)
		}
	}

	result := &SandboxResult{
		Success:         exitCode == 0,
		ExitCode:        exitCode,
		Stdout:          stdout.String(),
		Stderr:          stderr.String(),
		DurationSeconds: duration,
	}

	status := ExecutionCompleted
	if exitCode != 0 {
		status = ExecutionFailed
	}

	return &ExecutionHandle{
		ExecutionID: execID,
		AgentID:     agentID,
		SessionID:   sessionID,
		Status:      status,
		Result:      result,
	}, nil
}

// DestroySession force-removes the container for the given session.
func (p *DockerSandboxProvider) DestroySession(agentID, sessionID string) error {
	p.mu.Lock()
	key := containerKey(agentID, sessionID)
	name, ok := p.containers[key]
	if ok {
		delete(p.containers, key)
	}
	p.mu.Unlock()

	if !ok {
		return fmt.Errorf("session %s:%s not found", agentID, sessionID)
	}

	ctx, cancel := context.WithTimeout(context.Background(), dockerRemoveTimeout)
	defer cancel()
	out, err := exec.CommandContext(ctx, "docker", "rm", "-f", name).CombinedOutput()
	if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			return fmt.Errorf("docker rm timed out after %s", dockerRemoveTimeout)
		}
		return fmt.Errorf("docker rm failed: %w: %s", err, strings.TrimSpace(string(out)))
	}
	return nil
}

File: agent-governance-typescript/src/sandbox.ts
// Copyright (c) Microsoft Corporation. Licensed under the MIT License.

import { execSync, exec, execFileSync, execFile } from 'child_process';
import { randomUUID } from 'crypto';

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export enum SessionStatus {
  Provisioning = 'provisioning',
  Ready = 'ready',
  Executing = 'executing',
  Destroying = 'destroying',
  Destroyed = 'destroyed',
  Failed = 'failed',
}

export enum ExecutionStatus {
  Pending = 'pending',
  Running = 'running',
  Completed = 'completed',
  Cancelled = 'cancelled',
  Failed = 'failed',
}

// ---------------------------------------------------------------------------
// Data types
// ---------------------------------------------------------------------------

export interface SandboxConfig {
  timeoutSeconds: number;
  memoryMb: number;
  cpuLimit: number;
  networkEnabled: boolean;
  readOnlyFs: boolean;
  envVars: Record<string, string>;
}

export interface SandboxResult {
  success: boolean;
  exitCode: number;
  stdout: string;
  stderr: string;
  durationSeconds: number;
  killed: boolean;
  killReason: string;
}

export interface SessionHandle {
  agentId: string;
  sessionId: string;
  status: SessionStatus;
}

export interface ExecutionHandle {
  executionId: string;
  agentId: string;
  sessionId: string;
  status: ExecutionStatus;
  result?: SandboxResult;
}

// ---------------------------------------------------------------------------
// Default config factory
// ---------------------------------------------------------------------------

export function defaultSandboxConfig(): SandboxConfig {
  return {
    timeoutSeconds: 60,
    memoryMb: 512,
    cpuLimit: 1.0,
    networkEnabled: false,
    readOnlyFs: true,
    envVars: {},
  };
}

// ---------------------------------------------------------------------------
// Abstract interface
// ---------------------------------------------------------------------------

export interface SandboxProvider {
  createSession(agentId: string, config?: SandboxConfig): Promise<SessionHandle>;
  executeCode(agentId: string, sessionId: string, code: string): Promise<ExecutionHandle>;
  destroySession(agentId: string, sessionId: string): Promise<void>;
  isAvailable(): Promise<boolean>;
}

// ---------------------------------------------------------------------------
// Environment variables blocked from sandbox containers
// ---------------------------------------------------------------------------

const BLOCKED_ENV_VARS = new Set([
  'LD_PRELOAD',
  'LD_LIBRARY_PATH',
  'LD_AUDIT',
  'LD_DEBUG',
  'LD_PROFILE',
  'LD_SHOW_AUXV',
  'LD_DYNAMIC_WEAK',
  'PYTHONSTARTUP',
  'PYTHONPATH',
]);

function sanitizeEnvVars(envVars: Record<string, string>): Record<string, string> {
  const clean: Record<string, string> = {};
  for (const [k, v] of Object.entries(envVars)) {
    if (!BLOCKED_ENV_VARS.has(k.toUpperCase())) {
      clean[k] = v;
    }
  }
  return clean;
}

// Docker resource-name pattern: [a-zA-Z0-9][a-zA-Z0-9_.-]{0,127}
const DOCKER_NAME_RE = /^[a-zA-Z0-9][a-zA-Z0-9_.-]{0,127}$/;

function validateResourceName(value: string, label: string): void {
  if (!DOCKER_NAME_RE.test(value)) {
    throw new Error(
      `Invalid ${label} '${value}': must match [a-zA-Z0-9][a-zA-Z0-9_.-]{0,127}`,
    );
  }
}

// ---------------------------------------------------------------------------
// DockerSandboxProvider
// ---------------------------------------------------------------------------

export class DockerSandboxProvider implements SandboxProvider {
  private readonly image: string;
  private readonly containers = new Map<string, string>(); // sessionId -> containerId

  constructor(image: string = 'python:3.11-slim') {
    this.image = image;
  }

  async isAvailable(): Promise<boolean> {
    try {
      execSync('docker info', { stdio: 'pipe', timeout: 10_000 });
      return true;
    } catch {
      return false;
    }
  }

  async createSession(
    agentId: string,
    config?: SandboxConfig,
  ): Promise<SessionHandle> {
    validateResourceName(agentId, 'agentId');

    const cfg = config ?? defaultSandboxConfig();
    const sessionId = randomUUID();
    const containerName = `agt-${agentId}-${sessionId.slice(0, 8)}`;

    const args: string[] = [
      'run', '-d',
      '--name', containerName,
      '--cap-drop=ALL',
      '--security-opt=no-new-privileges',
      `--memory=${cfg.memoryMb}m`,
      `--cpus=${cfg.cpuLimit}`,
      '--pids-limit=256',
    ];

    if (cfg.readOnlyFs) {
      args.push('--read-only');
    }

    if (!cfg.networkEnabled) {
      args.push('--network=none');
    }

    const safeEnv = sanitizeEnvVars(cfg.envVars);
    for (const [k, v] of Object.entries(safeEnv)) {
      args.push('-e', `${k}=${v}`);
    }

    args.push(this.image, 'tail', '-f', '/dev/null');

    try {
      const containerId = execFileSync('docker', args, {
        stdio: 'pipe',
        timeout: 30_000,
      })
        .toString()
        .trim();

      this.containers.set(sessionId, containerId);

      return {
        agentId,
        sessionId,
        status: SessionStatus.Ready,
      };
    } catch (err: unknown) {
      const message = err instanceof Error ? err.message : String(err);
      throw new Error(`Failed to create sandbox session: ${message}`);
    }
  }

  async executeCode(
    agentId: string,
    sessionId: string,
    code: string,
  ): Promise<ExecutionHandle> {
    validateResourceName(agentId, 'agentId');

    const containerId = this.containers.get(sessionId);
    if (!containerId) {
      throw new Error(`No active session '${sessionId}' for agent '${agentId}'`);
    }

    const executionId = randomUUID();
    const startTime = Date.now();

    return new Promise<ExecutionHandle>((resolve) => {
      const encoded = Buffer.from(code).toString('base64');
      const execArgs = [
        'exec', containerId, 'python3', '-c',
        `import base64; exec(base64.b64decode('${encoded}').decode())`,
      ];

      execFile('docker', execArgs, { timeout: 60_000 }, (error, stdout, stderr) => {
        const durationSeconds = (Date.now() - startTime) / 1000.0;
        // Node's ExecException.code can be: a numeric exit code (child exited
        // non-zero), `null` (child killed by a signal — `error.signal` is set
        // instead), or a string like 'ENOENT' (the spawn itself failed). The
        // previous `error.code as number ?? 1` cast handled the `null` case
        // by accident — `null ?? 1` is `1` — but on the spawn-failure path it
        // let the string ('ENOENT') flow through as a `number`-typed exit
        // code, and downstream consumers treating `exitCode` as numeric saw
        // a string. Narrow explicitly: only accept a numeric code; otherwise
        // synthesise 1 for any error (signal kill, spawn failure, etc.).
        const exitCode = error
          ? typeof error.code === 'number' ? error.code : 1
          : 0;
        const killed = error !== null && 'killed' in error && (error as { killed: boolean }).killed;

        const result: SandboxResult = {
          success: exitCode === 0,
          exitCode,
          stdout: stdout ?? '',
          stderr: stderr ?? '',
          durationSeconds,
          killed,
          killReason: killed ? 'timeout' : '',
        };

        resolve({
          executionId,
          agentId,
          sessionId,
          status: exitCode === 0 ? ExecutionStatus.Completed : ExecutionStatus.Failed,
          result,
        });
      });
    });
  }

  async destroySession(agentId: string, sessionId: string): Promise<void> {
    validateResourceName(agentId, 'agentId');

    const containerId = this.containers.get(sessionId);
    if (!containerId) {
      return; // already destroyed or never existed
    }

    try {
      execFileSync('docker', ['rm', '-f', containerId], {
        stdio: 'pipe',
        timeout: 15_000,
      });
    } finally {
      this.containers.delete(sessionId);
    }
  }
}
